Adds require_* template type traits

SteveBronder opened the pull request:
## Summary


Out of the 2.8K of line additions 1.6K are tests. Most of the other changes are changing `enable_if` to `require`. The only files I would say have meaningful changes/additions are

1. (new) [`is_matrix_cl`](https://github.com/stan-dev/math/pull/1344/files#diff-7b389a0f140493a48e5db0055cc93948)
2. (change) [`matrix_cl`](https://github.com/stan-dev/math/pull/1344/files#diff-2336890ba7053d4cb33bf81074bb1a61) is used as a test bed to show some of the features below
3. (new) [`matrix_cl scalar_type`](https://github.com/stan-dev/math/pull/1344/files#diff-41128f4cc82232686988ce9bc2f5a323) gives `scalar_type_t` for `matrix_cl`
4. (new) [`matrix_cl value_type`](https://github.com/stan-dev/math/pull/1344/files#diff-e921226ee056286fc0848ba21890f793) gives `value_type_t` for matrix_cl
5. (new) [`require_generics.hpp`](https://github.com/stan-dev/math/pull/1344/files#diff-7286296d00c9b4a679f273d9c1d7ba69) Holds all of the `require` aliases explained below


Adds higher order functions for enabling and disabling functions through type traits. This new type system is what I'm going to call "legacy concepts" aka I think the closest thing you can get to C++20 concepts without having the `requires` keyword.

Legacy concepts are done through `enable_if_t` aliases which are called `require_*`. There are 5 basic types of requires.

1. `requires_*_t`: This means the function requires a certain type to turn on. For instance

```cpp
template <typename T, require_arithmetic_t<T>...>
T add(T x, T y) { return x + y; }
```

2. `require_not_*_t` : When we want to disable for a type

```cpp
// Only works for non-arithmetic types
template <typename T, require_not_arithmetic_t<T>...>
T add(T x, T y) { return x + y; }
```

3. `require_all_*_t` : Takes a parameter pack of types to enable if all types satisfy the check.

```cpp
// Works if T1 and T2 are arithmetic
template <typename T1, typename T2, require_all_arithmetic_t<T1, T2>...>
auto add(T1 x, T2 y) { return x + y; }
```

4. `require_any_*_t` : Takes a parameter pack of types to enable if any of the types satisfy the check.

```cpp
// Works if either T1 or T2 are arithmetic
template <typename T1, typename T2, require_any_arithmetic_t<T1, T2>...>
auto add(T1 x, T2 y) { return x + y; }
```

5. `require_not_any_*_t` : Takes a parameter pack of types to enable if any one of the types are not satisfied.

```cpp
// Works if either neither T1 or T2 are arithmetic
template <typename T1, typename T2, require_not_any_arithmetic_t<T1, T2>...>
auto add(T1 x, T2 y) { return x + y; }
```

6. `require_not_all_*_t` : Takes a parameter pack of types to enable if all of the types are not satisfied.

```cpp
// Works if neither T1 and T2 are arithmetic
template <typename T1, typename T2, require_not_all_arithmetic_t<T1, T2>...>
auto add(T1 x, T2 y) { return x + y; }
```

The above are built out for the meta programs to detect

```cpp
is_arithmetic
is_floating_point
is_double_or_int
is_same
is_var
is_var_or_arithmetic
is_fvar
is_var_or_fvar
is_stan_scalar (either var, fvar, or arithmetic)
is_std_vector
is_vector
is_vector_like
is_eigen
is_eigen_vector
```

in addition, `std::vector` and `Eigen` types have additional `requires` to detect if the `value_type` (the first underlying type) or the `scalar_type` (the containers underlying scalar type) satisfy a condition to enable a class or function. 

The container `requires` have a `_vt` and `_st` to symbolize the above. A function that accepts eigen matrices of whose whose value type is floating point types can be defined as 

```cpp
template <typename Mat1, typename Mat2,
  require_all_eigen_vt<is_floating_point, Mat1, Mat2>...>
auto add(Mat1&& A, Mat2&& B) { return A + B;}
```

A function that accepts standard vectors of Eigen vectors whose scalar type is floating point types can be defined as 

```cpp
template <typename Vec1, typename Vec2,
  require_all_std_vector_vt<is_eigen_vector, Vec1, Vec2>...,
  require_all_std_vector_st<is_floating_point, Vec1, Vec2>...>
auto add(Vec1&& A, Vec2&& B) { 
  std::vector<decltype<A[0] + B[0]>> return_vec;
  std::transform(A.begin(), A.end(), B.begin(), return_vec.begin(), 
    [](auto&& x, auto&& y) {
        return x + y;
    });
  return return_vec;
}
```


## Background
I'd like to start using perfect forwarding semantics in the math library. There's a few reasons for this

1. A lot of folks (including me) are looking to make Stan more performant through parallelization, but I think there is still a lot of money on the table when it comes to how we move and manage our memory in single threaded execution. 

2. If we want to add sparse matrices we either have to use more general templates in functions that can take in Eigen matrices or write overloads specializing on those. So for addition we would need signatures for adding sparse-sparse, sparse-dense, dense-sparse, and dense-dense matrices. That's not super fun, but the good news is that with more general templates (as perfect forwarding demands) we can support sparse matrices while getting better performance on the things we do now.

3. When using OpenCL with a cpu, if the Eigen type or vector coming in is a non const rvalue, we can pretty safely steal that memory to do a 'zero cost copy' and not pay any copy cost while getting the multicore performance.

You can see a blog post by Meyer's [here](https://isocpp.org/blog/2012/11/universal-references-in-c11-scott-meyers) that goes in depth on what perfect forwarding means and how to implement it I'll give a tl;dr below. But the tl;dr of the tl'dr is that using perfect forwarding semantics allow rvalues to be moved instead of copied to lvalues in the case of `const Blah&` in signatures, but perfect forwarding is very hard to implement correctly.

This PR will make signatures like this

```cpp
template <typename T1, typename T2, int R, int C>
inline Eigen::Matrix<return_type_t<T1, T2>, R, C> a_func(
    const Eigen::Matrix<T1, R, C>& m1, const Eigen::Matrix<T2, R, C>& m2) {
   check_nan(m1);
   check_nan(m2);
   Eigen::Matrix<return_type_t<T1, T2>, -1, -1> B = another_func(m1, m2);
    return B;
}
```

Look like this

```cpp
template <typename Mat1, typename Mat2, 
require_all_eigen_vt<is_arithmetic, Mat1, Mat2>...> // (1)
inline auto a_func(Mat1&& m1, Mat2&& m2) { // (2)
   check_nan(m1);
   check_nan(m2);
   // If m1 and/or m2 is an rvalue it will be moved over to this function
  // instead of copied to an lvalue
   auto B = another_func(std::forward<Mat1>(m1), std::forward<Mat2>(m2)); // (3)
   return B;
```

Where

(1) `require_*_eigen_vt` is a template template alias around `std::enable_if_t` that checks whether each type is both derived from the `EigenBase` class and has an underlying arithmetic value type. There are two `requires_*` for each container, `require_*_vt` and `require_*_st` where `vt` stands for `value_type` and `st` stands for `scalar_type`. When methods use containers of containers we want to verify both the underlying container is the correct type and the scalar type of the container is the correct type. For example, if we accept a vector of eigen vectors with arithmetic scalars for a certain function this would have the template.

```cpp
template <typename Vec, require_std_vector_vt<is_eigen, Vec>..., require_std_vector_st<std::is_arithmetic, Vec>...>
auto my_func(Vec&& v) {
/// do stuff
}
```

(2) The `...` comes from [this](https://www.fluentcpp.com/2019/08/23/how-to-make-sfinae-pretty-and-robust/) blog post and may seem a bit odd. But it's a clever way to force the requirement type trait to execute through parameter pack expansion. In the case of success the parameter pack ` require_all_eigen_t<condition, Mat1, Mat2>...`  has nothing in it and so nothing happens, but because we forced execution, failure will SFINAE out the signature for this function.

(3) In the signature the types only have to satisfy the requirements of (1), meaning that they can be Eigen matrices, expressions of eigen matrices, const/nonconst, and lvalues or rvalue refererences. The && in the signature acts like an `or` statement with lvalue (&) being true rvalue (&&) being false. If the type coming in is an lvalue reference then the type signature does `Mat1&& || &` produces an `Mat1&`. If the type coming in is an rvalue then the type signature does `Mat1&& || &&` and produces `Mat1&&`.

One other important note for this, while in the process of implimenting this, in functions like `add` that take in two Eigen expressions, we will have to surround the return statements with `(return_type).eval()` and also perform an `.eval()` before coefficient access. The `.eval()` on return is necessary for now since all the other math library functions can only take in Eigen matrices. Once all the signatures in the library can take in general Eigen types we can take off the training wheels and get a lot of benefit from Eigen's buildup of expression templates though the stan math functions.

(3) The `std::forward` in the call of `another_func` allows the matrices to keep their value type as they are passed to the next function. It's important to note that `std::forward` is used at the last call site before the end of the function. If `m1` was an rvalue and we used `std::forward` in `check_nan` then by the time we tried using it for `another_func` the memory would have been moved and the program would seg fault.

So that's pretty cool, we work with pretty large objects so not having to make a copy in each signature is good! There's some bureaucracy around the process but overall things are not too rough.
 


Out of the 2300 new lines in this pr, 1500 lines of this PR are testing, while out of the other 800 or so I'd say only 150 lines are actually meaningful. The important lines are at the top of [`require_generics.hpp`](https://github.com/stan-dev/math/compare/feature/require-templates?expand=1#diff-7286296d00c9b4a679f273d9c1d7ba69R23) where we define our base aliases that are used in the rest of the PR.

`require_t` is an alias for `enable_if_t` that takes in a type trait with a static member named `value`. it will enable or disable a signature depending on whether `value` is true or false. The rest of those base aliases are just clever manipulations of that. `require_not_t` being `std::enable_if_t<!...>`. `require_all_t` and ``require_any_t`` are variadic templated aliases for enable if that use `conjunction<...>` and `disjunction<...>` to turn on and off signatures depending on whether all (logical and) or any (logical or) of the type traits are satisfied, respectively.

The weirdest addition in this PR is the requirement type traits for containers which I'll paste here and then explain. In order to write something simplish like `require_any_vector_t<is_var, Var1, Var2>`  we need the following

```cpp
template <template <class...> class ContainerCheck, template <class...> class TypeCheck, class... Check>
 using container_type_check_base = bool_constant<math::conjunction<
                             ContainerCheck<std::decay_t<Check>>...,
                             TypeCheck<value_type_t<Check>>...>::value>;


 template <template <class...> class TypeCheck, class... Check>
 struct is_std_vector_check : container_type_check_base<is_std_vector, TypeCheck, Check...> {};

 template <template <class...> class TypeCheck, class... Check>
 using require_std_vector_t = require_t<is_std_vector_check<TypeCheck, Check...>>;
```

`container_type_check_base` is a `template <template template>`, this means it takes in two types that will be templated in the alias. So here both `ContainerCheck` and `TypeCheck` use the the template type `Check` when they are defined in the alias. The type `TypeCheck` uses `value_type_t` to extract the underlying type from the variadic template types `Check`.  `Check` is a variadic template type mostly to deal with when either `ContainerCheck` or `TypeCheck` have more than one template type in their arguments, though in most cases this is usually something of the form

```cpp
template <typename T, typename = void>
```
which is used frequently for partial specializations for type traits.

the struct alias `is_std_vector_check` is a partial specialization of `container_type_check_base`, 

```cpp
 template <template <class...> class TypeCheck, class... Check>
 struct is_std_vector_check : container_type_check_base<is_std_vector, TypeCheck, Check...> {};
```

It takes in only one templated template type and sets the first as `is_std_vector`. We then finish things off by making the final alias that gives us our nice pretty requirement

```cpp
 template <template <class...> class TypeCheck, class... Check>
 using require_std_vector_t = require_t<is_std_vector_check<TypeCheck, Check...>>;
```

And voila! We can now hide away all the dirt while using the above to make a bunch of pretty `requires` methods, which tbh are just copy/pasted versions of `require_std_vector`

To example what this looks like I've added the `requires` and perfect forwarding templates to the [`matrix_cl`](https://github.com/stan-dev/math/pull/1344/files#diff-2336890ba7053d4cb33bf81074bb1a61) constructor.

Edit: Sadly this won't work yet on the stuff in Stan math since the Stan upstream library has very fixed signatures. So this PR will need to happen in math, then I'll have to jump over to the Stan repo and implement the new signature style there first 

## Tests



Tests exist for checking all of the higher order functions included in `require_generics.hpp` files and can be tested with

```bash
./runTests.py ./test/unit/math -f require
```

These tests use the utilities in `test/unit/math/require_utils.hpp` which contain helper structs to define testing for unary and variadic `requires` aliases as well as larger structs containing the actual google tests used for the individual tests.

## Side Effects

value type has also been moved from the math namespace to the stan namespace where the rest of the type traits exist.

## Checklist

- [x] Math issue #1308

- [x] Copyright holder: Steve Bronder

- [x] the basic tests are passing

    - unit tests pass (to run, use: `./runTests.py test/unit`)
    - header checks pass, (`make test-headers`)
    - docs build, (`make doxygen`)
    - code passes the built in [C++ standards](https://github.com/stan-dev/stan/wiki/Code-Quality) checks (`make cpplint`)

- [x] the code is written in idiomatic C++ (lol) and changes are documented in the doxygen

- [x] the new changes are tested

SteveBronder commented:
Alright so all the tests are passing so I think this is ready for a general discussion.

Should this be broken up into multiple PRs? The new code is only about 500 lines while the rest are tests that are pretty copy/pasta

SteveBronder commented:
@syclik I'm tagging you because I'm not to sure how to move forward

bob-carpenter commented:
Thanks so much for the amazingly well documented PR.

I agree it's important to eliminate copying and believe this is the right, modern C++ way to do it. 

I'm pretty sure we don't have any pointer types anywhere where we'd use `scalar_type` to extract their base scalar types, but we'd definitely want it to keep going through pointers, so that could be fixed in an independent PR or included with this one if it's necessary.

Does `is_arithmetic` include our autodiff types now or was the first example meant to only apply to primitives?

P.S.  That's "et voila" when you get to France :-)

SteveBronder commented:
> Thanks so much for the amazingly well documented PR.
> I agree it's important to eliminate copying and believe this is the right, modern C++ way to do it.

Thanks!! Before the PR is merged I'll probably add this PRs summary to either the docs or the wiki.

> I'm pretty sure we don't have any pointer types anywhere where we'd use scalar_type to extract their base scalar types, but we'd definitely want it to keep going through pointers, so that could be fixed in an independent PR or included with this one if it's necessary.

I can make this in a separate PR

> Does is_arithmetic include our autodiff types now or was the first example meant to only apply to primitives?

It's only to primitives. `is_stan_scalar` is the one that works on all of our types while `is_ad_type` would work only on `fvar`s and `var`s.

I still can't decide whether we would want a `require_eigen<Mat1>` type trait that generally accepts any eigen container. With how non-informative the signatures will be I think `require_eigen_t<is_stan_scalar, Mat1>`(1) gives actually useful info relative to `require_eigen<Mat1>`(2).

(2) removes the `_t` since in my mind palace the `_t` means "an eigen container with type `blah`. If we are generally checking it's an eigen container I think it's good to remove that.

> P.S. That's "et voila" when you get to France :-)

I'll be caché'ing as we et voila across Paris ;-)

SteveBronder commented:
@t4c1 would you mind taking a look over this? Mostly a sanity check, I still need to figure out some way to automate making the tests for this so the line count additions do not look so ridiculous. The meaningful changes happened in


1. stan/math/opencl/is_matrix_cl.hpp (added a new trait)
2. stan/math/opencl/matrix_cl.hpp (changed constructors)
3. stan/math/prim/scal/meta/scalar_seq_view.hpp (uses pf)
   - This one looks goofy, `scalar_seq_view` is initailized with <>. It's tempalted type and constructor argument can differ in value type so we have another template on the constructor.
4. stan/math/prim/scal/prob/normal_cdf.hpp (just to make sure I'm doing pf stuff right)
5. stan/math/prim/scal/meta/require_generics.hpp (mostly what I mentioned above)

rok-cesnovar commented on stan/math/prim/scal/prob/normal_log.hpp:
```diff
@@ -30,7 +30,7 @@ namespace math {
  */
 template <bool propto, typename T_y, typename T_loc, typename T_scale>
-inline auto normal_log(T_y&& y, T_loc&& mu, T_scale&& sigma) {
+inline auto normal_log(T_y&& y, T_loc&& mu, T_scale&& sigma) { // NOLINT
```
Could we add this && case to the linter settings or something. This is so annoying.

SteveBronder replied:
Yeah I'm looking into that. Also editing my stuff above. I need to pull out the strict specialization in operands and partials

t4c1 commented:
I will look this over tomorrow.

SteveBronder commented:
> Since I'm not 100% sure and you have doubts I can make a benchmark in attempt to find out if there is any difference.

Yeah I def need to sit down hopefully this weekend and workout a benchmark over different sizes of Eigen matrices / vectors

SteveBronder commented on stan/math/prim/scal/meta/require_generics.hpp:
```diff
@@ -0,0 +1,476 @@
+ * Returns a type void if all of the conditions are false.
+ */
+template <class... Checks>
+using require_not_all_t
```
Yeah the verbage on this was weird for me and I kept flipping back and forth. I'll fix this up

SteveBronder commented on stan/math/opencl/matrix_cl.hpp:
```diff
@@ -234,13 +232,13 @@ class matrix_cl<T, enable_if_arithmetic<T>> {
-  explicit matrix_cl(const std::vector<Eigen::Matrix<T, R, C>>& A) try
-      : rows_(A.empty() ? 0 : A[0].size()),
-        cols_(A.size()) {
+  template <typename Vec, require_std_vector_t<is_eigen, Vec>...>
```
Containers of containers is a thing I need to work on. There's quite a few places where we have vectors of eigen matrices that then have a particular underlying scalar type. Not sure what is nicest syntax wise tho'

```
require_std_vector_contains_t<is_eigen, is_arithmetic, Vec>
```

I think I can do something like that

SteveBronder replied:
Thinking about this, it may be better to do two requires here instead of having one joint on (otherwise combinatorally I think the actual code would start looking weird and jankey.) How do you feel about 
```
template <typename Vec, require_std_vector_vt<is_eigen, Vec>..., require_std_vector_st<is_arithmetic, Vec>...>
```

Where `_vt` would be equivalent to doing `value_type` while `_st` would be looking at `scalar_type`?

t4c1 replied:
That is fine with me. Thinking about it I have another idea that may be clearer and a bit more general. 

``` template<typename Vec, require_std_vector<Vec>, require_eigen<contained_t<Vec>>, require_arithmetic<contained_t<contained_t<Vec>>>```

`contained_t` is generalization for `std::vector<T>::value_type` or `Eigen::Matrix<...>::Scalar`.

If you prefere yours that is also fine with me.

SteveBronder replied:
For yours we can use `value_type_t`. I have mixed feelings on the verbosity of it, though I think usually we either want to check

1. The containers underlying container (idt we have containers of containers of containers? Can we think of a situation where we explicitly have that?)
2. The containers underlying scalar type (var, fvar, arithmetic, yada yada)

I'll play around with both to see which is nicer

SteveBronder commented on stan/math/opencl/matrix_cl.hpp:
```diff
@@ -213,14 +211,14 @@ class matrix_cl<T, enable_if_arithmetic<T>> {
-      : buffer_cl_(A.buffer_cl_),
+  template <typename Mat, require_matrix_cl_t<std::is_arithmetic, Mat>...>
+  explicit matrix_cl(Mat&& A)
+      : buffer_cl_(std::forward<decltype(A.buffer_cl_)>(A.buffer_cl_)),
```
Dumb Q, if A is an rvalue are it's members still lvalues because they are named?

t4c1 replied:
Not sure. My gut feeling says they should be rvalues too.

SteveBronder replied:
> These std::forward calls do the same as std::move would, since buffer_cl_, write_events_ and read_events_ are not templated.

They are the same as std::move if A is an rvalue, if it's an lvalue it won't move (reference collapsing)

SteveBronder replied:
I think we are agreeing with what the code is doing but not the implications

I changed this here is to it handle both const and non-const lvalues and r values, is there a reason we want a separate constructor for all of them?

SteveBronder replied:
Gaaahhh nah nvm I'm dumb and your right we need to explicitly run that `clCopyBuffer`

SteveBronder commented on stan/math/opencl/matrix_cl.hpp:
```diff
@@ -308,10 +306,9 @@ class matrix_cl<T, enable_if_arithmetic<T>> {
-          A,
-      matrix_cl_view partial_view = matrix_cl_view::Entire)
+  template <typename Mat, require_eigen_t<std::is_arithmetic, Mat>...>
+  explicit matrix_cl(Mat&& A,
```
I want to preserve l/r value-ness along with const so for the bug we have right now we can use `is_r_value` to make a local copy when we move the data to the device. If the param coming in is not const we can give OpenCL the host pointer so on CPUs we don't have to make a copy

t4c1 replied:
> I want to preserve l/r value-ness along with const

Of what? I don't understand what you want to say here. This is not forwarding `A` to any other function. I don't think this change is a solution for that bug.

SteveBronder replied:
>> I want to preserve l/r value-ness along with const

> Of what? I don't understand what you want to say here. This is not forwarding A to any other function. 

It's not doing anything right now, but for zero copy you can't give OpenCL a const object's memory (because then that objects memory is gone violating constness). Having a generic template is going to let me in another PR look at both the valueness of the object (`A`) as well as it's constness so we can do things like

```
if (is_rvalue<Mat>::value) {
  // do a thing to fix the bug
} 
if (!is_const<Mat>::value) {
 // do the zero copy stuff
}

> I don't think this change is a solution for that bug.

Yeah not that I understand that bug more I agree. Though it lets us get to the solution

t4c1 replied:
Now I understand where you are going with this. Again if we want to support `Eigen::Map` or `Eigen::Block` without extra copies we need special handling. No need to do it in this PR.

SteveBronder commented on stan/math/opencl/matrix_cl.hpp:
```diff
@@ -322,7 +319,7 @@ class matrix_cl<T, enable_if_arithmetic<T>> {
       cl::Event transfer_event;
       queue.enqueueWriteBuffer(buffer_cl_, CL_FALSE, 0, sizeof(T) * A.size(),
-                               A.data(), nullptr, &transfer_event);
+                               A.eval().data(), nullptr, &transfer_event);
```
Indeed! I think we always want to right now right? Else the bug you saw earlier. Maybe some of these thing should go in a separate PR

t4c1 replied:
These changes to `matrix_cl` should definitely go in a separate PR. And I think we only want to evaluate expressions that are not `Eigen::Ref` or `Eigen::Map`.

SteveBronder replied:
> Calling .eval() on Eigen::Ref or Eigen::Map will make a copy.

I'm looking at the docs, If we map to a `Matrix` type won't it call `DenseBase::eval()` which doesn't make a copy?

> Notice that in the case of a plain matrix or vector (not an expression) this function just returns a const reference, in order to avoid a useless copy.

Or is a Map/Ref considered an expression?

https://eigen.tuxfamily.org/dox/classEigen_1_1DenseBase.html#aa73e57a2f0f7cfcb4ad4d55ea0b6414b

t4c1 replied:
Since this is templated now we won't get `Eigen::Ref` anyway. I guess supporting `Eigen::Map` without copies can wait.

SteveBronder replied:
Did you see my comment above? Doesn't Map on a Matrix use DenseBase::eval() which does not make a copy when calling `.eval()`?

t4c1 replied:
eval on Matrix is a no-op. eval on Map will make a Matrix from it - a copy.

SteveBronder commented on stan/math/opencl/matrix_cl.hpp:
```diff
@@ -338,8 +335,8 @@ class matrix_cl<T, enable_if_arithmetic<T>> {
-  explicit matrix_cl(const std::vector<T>& A, const int& R, const int& C)
-      : rows_(R), cols_(C) {
+  template <typename Vec, require_std_vector_t<std::is_arithmetic, Vec>...>
+  explicit matrix_cl(Vec&& A, const int& R, const int& C) : rows_(R), cols_(C) {
```
Same as above where if A is non-const we give the matrix_cl the host pointer for CPU zero copy stuff

SteveBronder commented:
> [wrt const Mat& template] This is the version I think we should use.

Also I don't disagree with this!

SteveBronder commented:
> I'm almost sure there will be no temporary if the type of formal argument matches actual argument as is the case if the argument is templated. 

Made a little playground to mess with this stuff

https://godbolt.org/z/h1tJ3V

I think your right, but sometimes? It looks like const ref forces a construction with both the `int` and `copy` constructors***** at the end of the call to `func_c` while `func_f` just uses a move. That makes sense to me since the temporary is just getting moved to the end lvalue.


***** This is only true if you are returning back the input or (I think) a non-trivial member. If you make the return blahblah.numba + 10; those constructors all go away

t4c1 commented:
> I think your right, but sometimes? It looks like const ref forces a construction with both the int and copy constructors***** at the end of the call to func_c while func_f just uses a move. That makes sense to me since the temporary is just getting moved to the end lvalue.
***** This is only true if you are returning back the input or (I think) a non-trivial member. If you make the return blahblah.numba + 10; those constructors all go away

That makes sense. I guess we need perfect forwarding only if a function can return one of its arguments. In all other cases const reference will be fine.

SteveBronder commented on stan/math/opencl/matrix_cl.hpp:
```diff
@@ -213,14 +211,14 @@ class matrix_cl<T, enable_if_arithmetic<T>> {
-
-  matrix_cl(matrix_cl<T>&& A)
-      : buffer_cl_(A.buffer_cl_),
+  template <typename Mat, require_matrix_cl_t<std::is_arithmetic, Mat>...>
```
`matrix_cl(matrix_cl<T>&& A)` only allows construction from rvalues and this allows for construction from l and r value `matrix_cl`s

SteveBronder commented on stan/math/opencl/matrix_cl.hpp:
```diff
@@ -427,54 +424,47 @@ class matrix_cl<T, enable_if_arithmetic<T>> {
-    }
-    view_ = a.view();
+  template <typename Mat, require_matrix_cl_t<std::is_arithmetic, Mat>...>
+  matrix_cl<T>& operator=(Mat&& a) {
```
Just updated the code, if you check out the code we just have one assign (the generic template one)

t4c1 replied:
We need to do very different thing in copy and move assignment operators. You are essentially making a move here. But copy assignment should copy the buffer. While this would be doable in single function with some ifs I think I prefere separate copy and move assignment operators. 

I'm ok with removing assignment of matrix of different type (which did the wrong thing anyway).

SteveBronder commented on stan/math/opencl/is_matrix_cl.hpp:
```diff
@@ -0,0 +1,80 @@
+ */
+template <typename T>
+struct is_matrix_cl<
+    T, std::enable_if_t<internal::is_matrix_cl_impl<std::decay_t<T>>::value>>
```
We need the indirecton for the cv qualifers and valueness. So use the impl let it match to refs and const etc.

t4c1 commented on stan/math/opencl/matrix_cl.hpp:
```diff
@@ -260,7 +258,7 @@ class matrix_cl<T, enable_if_arithmetic<T>> {
       queue.enqueueWriteBuffer(
           buffer_cl_, CL_FALSE, sizeof(double) * offset_size,
-          sizeof(double) * rows_, A[i].data(), nullptr, &write_event);
+          sizeof(double) * rows_, A[i].eval().data(), nullptr, &write_event);
```
Thinking on this again we won't get `Eigen::Ref` here anyway. Supporting `Eigen::Map` or `Eigen::Block` would require special handling (no `.eval()`, which I'm fine with leaving out for now.)

SteveBronder replied:
I got rid of these .evals

t4c1 commented on stan/math/opencl/zeros.hpp:
```diff
@@ -26,15 +26,16 @@ namespace math {
                         this->rows(), this->cols(), matrix_view);
-} catch (const cl::Error& e) {
+}  // namespace math
+catch (const cl::Error& e) {
```
I think some autoformatting went wrong here.

t4c1 commented on stan/math/opencl/matrix_cl.hpp:
```diff
@@ -414,63 +412,43 @@ class matrix_cl<T, enable_if_arithmetic<T>> {
     rows_ = a.rows();
     cols_ = a.cols();
+    this->wait_for_read_write_events();
+    buffer_cl_ = a.buffer_cl_;
```
This is also not OK. Copies of `cl::Buffer` are shallow - they point to same underlying memory on GPU. To get an actual deep copy that call to `enqueueCopyBuffer()` can not be avoided.

SteveBronder replied:
`cl::Buffer` uses the `Wrapper` classes `operator=` (it uses `Memory::operator=` but that's just `Wrapper<cl_mem>::operator=`

```
    Wrapper<cl_type>& operator = (const Wrapper<cl_type>& rhs)
    {
        if (this != &rhs) {
            detail::errHandler(release(), __RELEASE_ERR);
            object_ = rhs.object_;
            detail::errHandler(retain(), __RETAIN_ERR);
        }
        return *this;
    }
```

is this a shallow copy? `object_` is just defined as

```
protected:
    cl_type object_;
```

which doesn't look like it would be shallow?

t4c1 replied:
Documentation for `cl::Buffer` https://github.khronos.org/OpenCL-CLHPP/classcl_1_1_buffer.html says:
> See Memory for details about copy semantics, etc.

Doc for Memory https://github.khronos.org/OpenCL-CLHPP/classcl_1_1_memory.html states:

> Copies of these objects are shallow, meaning that the copy will refer to the same underlying cl_mem as the original. For details, see clRetainMemObject() and clReleaseMemObject().

I guess that the object you are looking at is just a handle to underlying buffer.

SteveBronder replied:
Ahh so now I get what your saying. But do we need a deep copy here? That assignment is going to call clReleaseMemObject() but that only releases the memory if the reference count is zero. But since the memory is being assigned the reference count should be greater than zero right?

t4c1 replied:
Deleting buffer here is not a problem, since we have anouther matrix that uses it. But we could have two matrix_cl objects that have same underlying memory. Well, we (you indluded) decided we want it to behave this way [here](https://github.com/stan-dev/math/issues/1307).

SteveBronder replied:
Do the tests I added below for assignment cover what you are worried about? They check that a locally scoped matrix_cl's memory is correctly passed to an object defined outside of it's scope without deleting it. If those make sense then we can just use the one assign operator that uses forward

Feels silly not having these tests before!

https://github.com/stan-dev/math/pull/1344/files#diff-44c6cc026f22cf47e68bef5053262f4cR30

t4c1 replied:
As I said before deleting is NOT a problem. Having two matrices with the same buffer could be. If one is modifieds will be the other one.

SteveBronder replied:
Totally forgot about that PR thanks for linking it, I'll fix these up to do the deep copies. I'll also add a test that shows when the shallow copy situation can break bad

SteveBronder replied:
:+1 added a test for this so in the future if someone goes perfect forward loco they will be summarily stopped from goofing

SteveBronder commented:
I did the `_st` and `_vt` thing for checking container `value_type` vs `scalar_type`. I'm annoyed at how stupid these tests are and want to fix those so they are less garbage. 

@bob-carpenter do you know of some way we can make a little type testing system for this? Of the top of my head I think I could write some helper macro

SteveBronder commented:
Also updated top level comment with the `st` and `vt` definitions. Need to clean up the docs for these in the code as well

SteveBronder commented:
@t4c1 would you mind giving this a look tomorrow? Sorry to be slow on it, had to wrassle pretty hard to cover all the edge cases in the unit tests in a way that wouldn't cause 5K+ lines of testing code. 

I updated the top comment with info on which files in the math library have meaningful changes

t4c1 commented on stan/math/opencl/matrix_cl.hpp:
```diff
@@ -236,18 +235,20 @@ class matrix_cl<T, enable_if_arithmetic<T>> {
     // creates the OpenCL buffer to copy the Eigen
     // matrix to the OpenCL device
     buffer_cl_ = cl::Buffer(ctx, CL_MEM_READ_WRITE, sizeof(T) * size());
+    cl::Event write_event;
```
Why did you move this outside the loop? We need to add all write events, not just the last one.

SteveBronder replied:
Ah yes! So, do we? I think it depends on whether OpenCL does async when it writes to an individual buffer or if writes to a single buffer happen sequentially. Maybe there is not a guarantee around that. If I can find something in the docs that shows that is true then I'll revert this and else we can discuss it.

SteveBronder replied:
Yeah I think we do, fixed this

t4c1 commented on stan/math/opencl/matrix_cl.hpp:
```diff
@@ -259,12 +260,12 @@ class matrix_cl<T, enable_if_arithmetic<T>> {
+      queue.enqueueWriteBuffer(buffer_cl_, CL_FALSE,
+                               sizeof(double) * offset_size,
+                               sizeof(double) * rows_, A[i].data(),
+                               &this->read_events(), &write_event);
```
This is a constructor - this matrix was not used before. So it has no read events and this change is unnecessary.

t4c1 commented on stan/math/opencl/matrix_cl.hpp:
```diff
@@ -324,7 +324,7 @@ class matrix_cl<T, enable_if_arithmetic<T>> {
       cl::Event transfer_event;
       queue.enqueueWriteBuffer(buffer_cl_, CL_FALSE, 0, sizeof(T) * A.size(),
-                               A.data(), nullptr, &transfer_event);
+                               A.data(), &this->read_events(), &transfer_event);
```
Again there are no read events in a new matrix.

t4c1 commented on stan/math/opencl/matrix_cl.hpp:
```diff
@@ -347,8 +347,8 @@ class matrix_cl<T, enable_if_arithmetic<T>> {
-      queue.enqueueWriteBuffer(buffer_cl_, CL_FALSE, 0, sizeof(T), &A, NULL,
-                               &transfer_event);
+      queue.enqueueWriteBuffer(buffer_cl_, CL_FALSE, 0, sizeof(T), &A,
+                               &this->read_events(), &transfer_event);
```
No read events in a new matrix

t4c1 commented on stan/math/opencl/matrix_cl.hpp:
```diff
@@ -377,7 +378,7 @@ class matrix_cl<T, enable_if_arithmetic<T>> {
       cl::Event transfer_event;
       queue.enqueueWriteBuffer(buffer_cl_, CL_FALSE, 0, sizeof(T) * A.size(),
-                               A.data(), nullptr, &transfer_event);
+                               A.data(), &this->read_events(), &transfer_event);
```
Again

t4c1 commented on stan/math/opencl/matrix_cl.hpp:
```diff
@@ -394,7 +395,8 @@ class matrix_cl<T, enable_if_arithmetic<T>> {
    */
-  explicit matrix_cl(const double* A, const int& R, const int& C,
+  template <typename U, require_same_t<T, U>...>
+  explicit matrix_cl(const U* A, const int& R, const int& C,
```
Thanks for fixing this.

t4c1 commented on stan/math/opencl/matrix_cl.hpp:
```diff
@@ -406,7 +408,7 @@ class matrix_cl<T, enable_if_arithmetic<T>> {
       cl::Event transfer_event;
       queue.enqueueWriteBuffer(buffer_cl_, CL_FALSE, 0, sizeof(T) * size(), A,
-                               NULL, &transfer_event);
+                               &this->read_events(), &transfer_event);
```
Again

t4c1 commented on stan/math/opencl/matrix_cl.hpp:
```diff
@@ -416,63 +418,46 @@ class matrix_cl<T, enable_if_arithmetic<T>> {
+    this->cols_ = a.cols();
+    cl::Context& ctx = opencl_context.context();
+    cl::CommandQueue queue = opencl_context.queue();
+    try {
```
Why did you remove waiting for read and write events on this? I think we kinda need it.

SteveBronder replied:
I think we only need to wait for the write events in the copyBuffer function? Cuz reads in this context are fine, its just the writes we worry about

```cpp
      queue.enqueueCopyBuffer(a.buffer(), this->buffer(), 0, 0,
                              a.size() * sizeof(T), &a.write_events(),
                              &cstr_event);
```

t4c1 commented on stan/math/opencl/matrix_cl.hpp:
```diff
@@ -216,8 +215,8 @@ class matrix_cl<T, enable_if_arithmetic<T>> {
-  matrix_cl(matrix_cl<T>&& A)
-      : buffer_cl_(A.buffer_cl_),
+  explicit matrix_cl(matrix_cl<T>&& A)
+      : buffer_cl_(std::move(A.buffer_cl_)),
```
Thanks for fixing this.

syclik commented:
Dope!

I want to make sure the doc you wrote up here:
1. Doesn’t get lost
2. Is easy to find by other devs. 

I don’t think you need more doc, I just think it’d be a shame if it was buried in this PR and we’d have to find the PR to know what happened. 

Would it make sense to use something like [doygen’s modules](http://www.doxygen.nl/manual/grouping.html) to put your text into the codebase and have it render like this: https://mc-stan.org/math/modules.html ?

One question about naming (this is a question, not a suggestion or required change): you’ve got “not” mixed in with “or” like `require_not_double_or_int_t` and I recognize you’re negating the “double or int” condition. **Should this be “not double and not int” or maybe “not double nor int”?** In words, it’s hard to put parens, so it's read as “(not double) or (int)” instead of “not (double or int).” This is a weird edge case so it’s really a question. As long as it’s clearly documented, it’s ok (and we don’t have a foreseeable need for “(not double) or (int)” type of function).

SteveBronder commented on stan/math/prim/scal/meta/require_generics.hpp:
```diff
@@ -0,0 +1,769 @@
+
+template <typename T, typename... Types>
+using require_all_not_same_t
+    = require_all_not_t<std::is_same<std::decay_t<T>, std::decay_t<Types>>...>;
```
I think I flat out need to rethink how same works here. Reading it now it checks that the `Types` match with just `T`, but it really should check that each type is the same. If i figure that out I think this will make more sense / I can have an `any` for same as well

SteveBronder replied:
Actually I think just changing it to `any` makes sense (also i don't feel like writing out the more complicated thing which I'm not sure we would use that much)

SteveBronder commented:
> Dope!

Ty!

> I don’t think you need more doc, I just think it’d be a shame if it was buried in this PR and we’d have to find the PR to know what happened.

> Would it make sense to use something like doygen’s modules to put your text into the codebase and have it render like this: https://mc-stan.org/math/modules.html ?

Yes def, I've been thinking about that more and more, it would be nice to have modules for 

1. This
2. OpenCL
3. TBB
4. A shortened version of the math paper detailing how we do things

I'll look into this later today

> One question about naming (this is a question, not a suggestion or required change): you’ve got “not” mixed in with “or” like require_not_double_or_int_t and I recognize you’re negating the “double or int” condition. Should this be “not double and not int” or maybe “not double nor int”? In words, it’s hard to put parens, so it's read as “(not double) or (int)” instead of “not (double or int).” This is a weird edge case so it’s really a question. As long as it’s clearly documented, it’s ok (and we don’t have a foreseeable need for “(not double) or (int)” type of function).

Yeah it's a good Q, naming things is hard! I wanted to keep all of the names the same to avoid gotchas, but I def agree that `not_double_or_int` could be read in a lot of different ways. I'll think about this. It's not used anywhere so maybe the answer is to just remove it

SteveBronder commented:
@t4c1 Alrighty! I think I addressed all your review comments. I switched `require_all_not_same` to `require_any_not_same` since I think that makes more sense and removed the `this->read_events()` from the constructors. in addition I noticed that we commonly want to check for the same scalar or value type as a class template in our member functions so I added a `_st` and _vt` variant of `require_same`.

@syclik do you mind if I add the doxygen module in a separate PR? This one is already v big.

SteveBronder commented:
Alrighty! @t4c1 I did a look over the code and think I got all of your review comments and this is passing so if you want to have a look over I think it's good to go!

syclik commented:
Sorry -- missed the notification earlier. If it's not going into this PR,
then can you open a new issue with enough detail on what needs to be done?
Maybe someone else can tackle the doc or at the very least, it'll remind
you that there's outstanding work to be done on this PR.

On Tue, Oct 8, 2019 at 6:05 PM Steve Bronder <notifications@github.com>
wrote:

> Alrighty! @t4c1 <https://github.com/t4c1> I did a look over the code and
> think I got all of your review comments and this is passing so if you want
> to have a look over I think it's good to go!
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/stan-dev/math/pull/1344?email_source=notifications&email_token=AADH6FYVZHOO77U7QLVVR3LQNT723A5CNFSM4IUQ57SKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEAVYPAA#issuecomment-539723648>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AADH6F6NJLF7ECI5IMPOUULQNT723ANCNFSM4IUQ57SA>
> .
>

SteveBronder commented:
> Sorry -- missed the notification earlier. If it's not going into this PR,
then can you open a new issue with enough detail on what needs to be done?
Maybe someone else can tackle the doc or at the very least, it'll remind
you that there's outstanding work to be done on this PR.

Gonna do it today!
